src/python/pants/backend/python/rules/coverage.py

```diff
@@ -295,21 +297,28 @@ async def generate_coverage_reports(
             )
         )
     results = await MultiGet(Get(ProcessResult, PexProcess, process) for process in pex_processes)
+    result_snapshots = await MultiGet(Get(Snapshot, Digest, r.output_digest) for r in results)
+    result_stdouts = tuple(r.stdout for r in results)
     coverage_reports.extend(
-        _get_coverage_reports(coverage_subsystem.output_dir, report_types, results)
+        _get_coverage_reports(
+            coverage_subsystem.output_dir, report_types, result_stdouts, result_snapshots
+        )
     )
     return CoverageReports(tuple(coverage_reports))
 
 
 def _get_coverage_reports(
     output_dir: PurePath,
     report_types: Sequence[CoverageReportType],
-    results: Tuple[ProcessResult, ...],
+    result_stdouts: Tuple[bytes, ...],
+    result_snapshots: Tuple[Snapshot, ...],
 ) -> List[CoverageReport]:
     coverage_reports: List[CoverageReport] = []
-    for result, report_type in zip(results, report_types):
+    for result_snapshot, result_stdout, report_type in zip(
+        result_snapshots, result_stdouts, report_types
+    ):
```

Comment:
This doesn't work great that `report_types`, `result_stdouts`, and `result_snapshots` are all expected to be the same length and to correspond to each other. That's not expressed explicitly via the function signature. Python won't complain if `zip` if the lengths are different (https://www.python.org/dev/peps/pep-0618/).

Instead, I recommend rewriting this function to return a single `CoverageReport`. Then, move the list into `generate_coverage_reports`, where you can set up the `zip` there, where it is more obvious that those collections are supposed to correspond.